process/track/metaBlockTrack.go

```diff
@@ -0,0 +1,130 @@
+		return nil, process.ErrNilHeadersNoncesDataPool
+	}
+
+	crossNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	crossNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	bbt := &baseBlockTrack{
+		hasher:                        arguments.Hasher,
+		headerValidator:               arguments.HeaderValidator,
+		marshalizer:                   arguments.Marshalizer,
+		rounder:                       arguments.Rounder,
+		shardCoordinator:              arguments.ShardCoordinator,
+		metaBlocksPool:                arguments.PoolsHolder.MetaBlocks(),
+		shardHeadersPool:              arguments.PoolsHolder.ShardHeaders(),
+		headersNoncesPool:             arguments.PoolsHolder.HeadersNonces(),
+		store:                         arguments.Store,
+		crossNotarizer:                crossNotarizer,
+		selfNotarizer:                 selfNotarizer,
+		crossNotarizedHeadersNotifier: crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier:  selfNotarizedHeadersNotifier,
+	}
+
+	err = bbt.initNotarizedHeaders(arguments.StartHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	mbt := metaBlockTrack{
+		baseBlockTrack: bbt,
+	}
+
+	blockProcessor, err := NewBlockProcessor(
+		arguments.HeaderValidator,
+		arguments.ShardCoordinator,
+		&mbt,
+		crossNotarizer,
+		crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mbt.blockProcessor = blockProcessor
+
+	mbt.headers = make(map[uint32]map[uint64][]*headerInfo)
+	mbt.metaBlocksPool.RegisterHandler(mbt.receivedMetaBlock)
+	mbt.shardHeadersPool.RegisterHandler(mbt.receivedShardHeader)
+
+	return &mbt, nil
+}
+
+func (mbt *metaBlockTrack) getSelfHeaders(headerHandler data.HeaderHandler) []*headerInfo {
+	selfMetaBlocksInfo := make([]*headerInfo, 0)
+
+	header, ok := headerHandler.(*block.Header)
+	if !ok {
+		log.Debug("getSelfHeaders", process.ErrWrongTypeAssertion)
+		return selfMetaBlocksInfo
+	}
+
+	for _, metaBlockHash := range header.MetaBlockHashes {
+		metaBlock, err := process.GetMetaHeader(metaBlockHash, mbt.metaBlocksPool, mbt.marshalizer, mbt.store)
+		if err != nil {
+			log.Debug("GetMetaHeader", err.Error())
+			continue
+		}
+
+		selfMetaBlocksInfo = append(selfMetaBlocksInfo, &headerInfo{hash: metaBlockHash, header: metaBlock})
+	}
+
+	return selfMetaBlocksInfo
+}
+
+func (mbt *metaBlockTrack) computeLongestSelfChain() (data.HeaderHandler, []byte, []data.HeaderHandler, [][]byte) {
+	lastSelfNotarizedHeader, lastSelfNotarizedHeaderHash, err := mbt.selfNotarizer.getLastNotarizedHeader(mbt.shardCoordinator.SelfId())
+	if err != nil {
+		return nil, nil, nil, nil
```

Comment:
masked error. i think it should be treated.

Reply:
...or at least logged.

Reply:
I have added a log. Actually getLastNotarizedHeader could not return err as it is initialized with startHeaders(genesis header for each shard) on New.. method.This method returns the new finalized headers constructed above the last finalized one in self chain. If the chain didn't advance and the node is sync or the last notarized header could not be determined, this method has nothing to return.

Reply:
These shard headers contained in the received meta block are used to notify all the registered handlers (for now only fork detector) that there are new self headers notarized in meta, so the node can now check sooner if it constructed on the notarized chain or not (not need anymore to wait until it will include these meta headers in its processed shard headers to determine this)